### DISCUSSION TITLE
`ChoiceField` is incompatible with `TextChoices`

### DISCUSSION THREAD

**Discussion by JulianOrteil** [encode/django-rest-framework#8965] [Potential Issue]
#8132 raised an issue in 2021 where `ChoiceField` has incorrect mappings. While I am unsure if the issue-raiser has encountered the same issue I am facing, this line of code is nonetheless causing problems for me.

Take the following `TextChoices` object. When we attempt to serialize it through a serializer, the validation error `"X" is not a valid choice` is thrown. This is because the value submitted to the serializer is the label, not the value.

```python
class ReadingType(models.TextChoices):
    humidity = "humidity", "Humidity"
    temperature = "temperature", "Temperature"
```

```python
class MySerializer(serializers.ModelSerializer):
    reading_type = serializers.ChoiceField(choices=ReadingType.choices)
    # ...
```

Diving into the cause, it can be seen that the following line--the one focused on in #8132--is the root cause. It fails to create a proper mapping between human-labels and database-values that exist in choice classes:

```python
print(self.choice_strings_to_values)
# {'humidity': 'humidity', 'temperature': 'temperature'}
# note database-value: database-value
```

I've written a small patch to make this work for my code which use both `TextChoices` and `IntegerChoices`:
```python
    def _set_choices(self, choices):
        self.grouped_choices = to_choices_dict(choices)
        self._choices = flatten_choices_dict(self.grouped_choices)

        # Map the string representation of choices to the underlying value.
        # Allows us to deal with eg. integer choices and text choices
        # while supporting either integer or string input, but still get
        # the correct datatype out.
        self.choice_strings_to_values = {}
        if isinstance(self.choices, OrderedDict):
            choice_container = self.choices.items()
        else:
            choice_container = self.choices
        for key, value in choice_container:
            if isinstance(value, str):
                self.choice_strings_to_values[value] = str(key)
            else:
                self.choice_strings_to_values[str(key)] = key
```

Which now outputs:
```python
print(self.choice_strings_to_values)
# {'Humidity': 'humidity', 'Temperature': 'temperature'}
# note human-label: database-value
```

**Comment by auvipy** [comment]
related https://github.com/encode/django-rest-framework/pull/8955 but need django built in feature support

**Comment by saadullahaleem** [comment]
Can you please check if #8968 solves this?
